Allow LDAP bind password to be stored in file, not database. Also, reduce number of calls to LDAP server for 2FA.

bpb54321 opened the pull request:
This pull request has two major changes.

1. Add an option for the plugin to grab the LDAP server bind password from an ldap.ini file in the plugin root directory. This addresses issue #1.
2. Removes a duplicate call to authenticate the user with the LDAP server. I believe this was initially in the code for the LDAP "optional" option, which allows the plugin to first try LDAP, then if LDAP doesn't authenticate, so authenticate using the default Omeka authentication. The problem that this causes is, if the LDAP server is set up with two-factor authenication (2FA), where the user has to approve a push notification on their phone before the authentication goes through, this forces the user to have to approve 4 total push notications, since the Zend LDAP adapter already authenticates the user twice with the LDAP server. I feel that approving two push notifications is the limit of patience for most people...approving four push notifications is really annoying. To my mind, we can avoid this annoyance by simply forcing people to choose LDAP auth is "required" or "disabled" in the plugin settings. If we do that, then we won't need the extra LDAP authentication call in the plugin.

I'm totally open to adjusting this pull request and making adjustments as needed. Thanks for being open to my changes...I hope to deploy this on a production site at Georgetown University!

bpb54321 committed:
Created a config option called central_auth_ldap_use_ini.

bpb54321 committed:
Ignoring ldap.ini in Git.

bpb54321 committed:
Get LDAP bind password from ini file.

bpb54321 committed:
Removed premptive authentication with LDAP before returning adapter.

Previously, the code authenticated using LDAP first to "try out" whether LDAP authentiation works before returning the LDAP adapter in the login filter. This is a problem if your university uses 2-factor authentication, because it doubles the number of 2-factor authentication push requests that must be responded to in order to log in. My opinion is that we should just keep it simple, and LDAP auth should either be "on" or "off".

bpb54321 committed:
Grab options from ldap.ini only if a bind username was entered in the options.

bpb54321 committed:
Remove 'optional' ldap option.

Since we've edited the code to only allow ldap to be either 'off' or 'on.

bpb54321 committed:
Updated composer.lock.

bpb54321 committed:
Revert "Updated composer.lock."

This reverts commit 76ef74fc01adc2b3dabf8a44f6a987a6538e9d13.

bpb54321 commented:
Hi @kloor, have you been able to take a look at the pull request that I submitted?

kloor commented:
Hello @bpb54321, and thank you for your pull request. 

With our spring semester beginning next week, I haven't had a chance to fully review this pull request yet. It is unlikely I'll merge the removal of marking LDAP as optional, as that is a feature we currently use to allow for both users with affiliation with our University and external volunteers to log in.

I thought I had initially tested that validation against LDAP would only occur once, and that the Zend authentication system should reuse the initial validation, though I'll look into testing again to see if that is so. I wasn't aware of any systems that use 2FA with LDAP. I'm curious, are you using the credentials from your own account within the plugin's configuration?

bpb54321 commented:
Hi John @kloor,

Thanks for your response, and sorry to pressure you in the beginning of the semester!
Here at Georgetown University, we have moved to Shibboleth as our primary SSO solution / method, but prior to that we used LDAP, and so we still maintain a legacy LDAP authentication server. Given that our university has transitioned all accounts to using 2FA, whenever a user with a Georgetown "netid" queries the LDAP server, they must authenticate the query with a push to an app on their smartphone.

I'm totally happy to continue to edit my changes so that LDAP auth can be an option that complements non-LDAP authentication as well, since that's a feature that you and I'm sure other folks use: I just did a quick-and-easy solution of making LDAP on/off to get the ball rolling with this pull request.

However, my understanding of how the plugin is structured is that the way you determine whether a user can authenticate via LDAP, or can only authenticate using the Omeka user system, is that you first test the user credentials on the LDAP server, and then if they fail to authenticate there, you use the regular Omeka authentication adapter. This makes total sense, but the problem with this is that it requires two 2FA pushes for the initial call, then two 2FA pushes for the actual call, if you are in fact able to authenticate with LDAP. Could we perhaps add a select field to the login screen so that the user could manually select whether or not they are logging in via LDAP or regular Omeka login?

I'll try to add some comments to the code itself to point out where the multiple authentications are happening. Look forward to working with you once your workload dies down!

bpb54321 commented on CentralAuthPlugin.php:
```diff
@@ -225,21 +226,30 @@ public function filterLoginAdapter($adapter, $args)
             );
 
-            // Attempt to authenticate using LDAP.
-            $result = $adapterLdap->authenticate();
```
@kloor Here is where the initial Ldap authentication happens, to check whether or not the user can use Ldap, or whether they must use the default Omeka authentication.

bpb54321 commented on CentralAuthPlugin.php:
```diff
@@ -225,21 +226,30 @@ public function filterLoginAdapter($adapter, $args)
-            // If the user authenticated or LDAP authentication is required,
-            // return the LDAP auth adapter as the version to use.
-            if ($result->isValid() || $ldap == 'required') {
-                return $adapterLdap;
```
@kloor This function actually returns an LDAP adapter to the Omeka authentication adapter filter, so the Omeka login system will use the Ldap adapter that was returned by the filter and authenticate with the LDAP server a second time. As a reminder, apparently in the Zend code, it authenticates twice with the ldap server if you use a bind with a username and password. This results in a total of 4 calls to the ldap server for my particular situation.

bpb54321 commented:
@kloor To answer your last question, in the plugin configuration settings, I check Bind Requires DN, fill in the username with the bind ldap account of my department, then the password with the bind password of my department.

kloor commented:
Yes, reviewing the code I see that the LDAP authentication would be run twice. My first thought for a fix that would still allow LDAP to be optional would be to modify `CentralAuth_LdapAdapter::authenticate()` to either store in a property the [result of the call to the parent method on line 28](https://github.com/BGSU-LITS/omeka-plugin-CentralAuth/blob/master/adapters/LdapAdapter.php#L28) or perhaps the result of all of the authentication attempts in the method. Then, the method could return on that property if it is specified instead of authenticating again.

Looking at the Zend code, it first it [binds with the user's credentials on line 315](https://github.com/omeka/Omeka/blob/master/application/libraries/Zend/Auth/Adapter/Ldap.php#L315
) every time, then [rebinds with the configuration credentials if they are specified on lines 324-327](https://github.com/omeka/Omeka/blob/master/application/libraries/Zend/Auth/Adapter/Ldap.php#L324-L327) and then [rebinds with the user's credentials again after their group membership is checked in lines 336-339](https://github.com/omeka/Omeka/blob/master/application/libraries/Zend/Auth/Adapter/Ldap.php#L336-L339). So, I see how that part can result in double 2FA requests. Probably the only way to work around that is to move all of the method's code into the `CentralAuth_LdapAdapter::authenticate()` method.

bpb54321 commented:
Hi John @kloor, thanks for your comment.

I like your idea of storing any previous authentication results as a property of the CentralAuth_LdapAdapter instance, and checking that before doing any further authentication. My team has determined that we can accept up to two 2FA push requests, so there's no need to redo the Zend Ldap adapter code, especially since I don't feel super knowledgable about the technical details about the whole LDAP process.

So I'll get to work on the change you suggested, and resubmit the pull request when it's ready. Thanks again!

bpb54321 committed:
Restored original plugin code which does an LDAP authentication if LDAP option is set to 'optional'.

bpb54321 committed:
Added $auth_result property on CentralAuth_LdapAdapter so reduce duplicate calls to LDAP server.

bpb54321 commented:
Hi John @kloor,

Please review my most recent commits when you get a chance. I implemented your idea of storing previous authentication results.

Let me know if further refinement is needed in the code.

-Brian

kloor committed:
Reduce changed lines and match code style.

I've pulled out the else statements that were added, as they are not necessary when the if statement returns and they made the differences hard to view.

I also changed the property name and visibility to match other Zend_Auth_Adapter code, and made the style of the documentation consistent.

kloor commented:
I think the changes to CentralAuth_LdapAdapter are good to go now.

Can you remove the changes to composer.lock? They are unrelated, so I don't want to include them when I squash and merge the PR.

For the ldap.ini file, is there a reason you named the setting "bind_password"? I'm also wondering if the ldap.ini should support all of the options passed to the adapter, so users could just copy the file between projects.

bpb54321 commented:
Hi @kloor , I reverted the changes to composer.lock.

And thank you for removing the else statements...I should've not tried to adjust the code style.

I named the setting "bind_password" because, from my limited understanding of how LDAP authentication works, we first need to "bind" to the LDAP server with my department's username and password, then verify that the user's username and password are correct. So I guess it's to distinguish it from the username and password of the user who is logging in. However, I'm open to you renaming it to whatever makes the most sense to you.

Regarding allowing ldap.ini to support all the LDAP options for the plugin, that makes sense to me! From a user experience point of view, should we keep the toggle on the config page which controls whether the options will be pulled from ldap.ini or from the database, or should we simply first look in the database, and then overwrite any database options that are found with any options that are found in ldap.ini? If we do the second approach, should we add a notice to this effect in the plugin configuration screen, or simply in the plugin Readme or description? Or both?

kloor commented:
To get this pull request wrapped up, I propose that the ini file be renamed `central_auth.ini`, the section can stay as `[ldap]` and the option be changed to just `password`. That way it would somewhat match the Omeka option of `central_auth_ldap_password` and also allow for adding the other LDAP options as well as SSO options in the future.

I think I like the idea of having the value in the ini file just take precedence over the database, since the configuration form for the plugin is already so long. I think its sufficient to have the documentation in the README.

bpb54321 committed:
Updated .gitignore to ignore central_auth.ini instead of ldap.ini.

bpb54321 committed:
Removed the option central_auth_ldap_use_ini.

bpb54321 committed:
Checking central_auth.ini for ldap options first, then fall back to DB.

bpb54321 committed:
Updated Readme with information on using central_auth.ini for ldap conf.

bpb54321 committed:
Added helper text to the ldap password filed in the config form.

bpb54321 commented:
@kloor I made the changes you suggested. Let me know what you think!

kloor committed:
Remove else and keep down number of changed lines.

kloor commented:
@bpb54321 Thanks for all the work. I have one unrelated change to commit, and then I'll package a new release.
